src/main/java/org/opentripplanner/routing/algorithm/transferoptimization/services/TransferGenerator.java

```diff
@@ -117,65 +117,102 @@ private List<TripToTripTransfer<T>> findAllTransfers(int stopPos) {
     return result;
   }
 
+  /**
+   * Find potential transfers where traveller does not have to walk between stops
+   * @param from initial trip
+   * @return list of potential transfers
+   */
   private Collection<TripToTripTransfer<T>> transferFromSameStop(TripStopTime<T> from) {
-    final int stop = from.stop();
-    var tx = transferServiceAdaptor.findTransfer(from, toTrip, stop);
-
-    if (tx != null && tx.getTransferConstraint().isNotAllowed()) {
-      return List.of();
-    }
+    var result = new ArrayList<TripToTripTransfer<T>>();
 
-    final int earliestDepartureTime = earliestDepartureTime(
-      from.time(),
-      SAME_STOP_TRANSFER_TIME,
-      tx
-    );
+    final int stop = from.stop();
 
-    final int toTripStopPos = toTrip.findDepartureStopPosition(earliestDepartureTime, stop);
+    // Find all possible transfers on given stop index, starting on from.time
+    var possibleTransfers = toTrip.findStopPositionsForStopIndex(from.time(), stop);
+
+    // Loop through transfers and decide whether they are possible
+    for (var stopPos : possibleTransfers) {
+      // Find transfer constraint for stop position
+      var tx = transferServiceAdaptor.findTransfer(
+        from,
+        toTrip,
+        stop,
+        from.stopPosition(),
+        stopPos
+      );
 
-    if (toTripStopPos < 0) {
-      return List.of();
-    }
+      if (!isAllowedTransfer(stopPos, tx)) {
+        continue;
+      }
 
-    boolean boardingPossible = toTrip.pattern().boardingPossibleAt(toTripStopPos);
+      // Check whether traveller will have enough time to do the transfer
+      // We have to do it here because every stop position may have unique transfer constraint
+      // So it may be possible to transfer at stop position 2 but not on 1...
+      final int earliestDepartureTime = earliestDepartureTime(
+        from.time(),
+        SAME_STOP_TRANSFER_TIME,
+        tx
+      );
+      if (earliestDepartureTime > toTrip.departure(stopPos)) {
+        continue;
+      }
 
-    if (!boardingPossible) {
-      return List.of();
+      // Add as a possible result
+      result.add(new TripToTripTransfer<>(from, TripStopTime.departure(toTrip, stopPos), null, tx));
     }
 
-    return List.of(
-      new TripToTripTransfer<>(from, TripStopTime.departure(toTrip, toTripStopPos), null, tx)
-    );
+    return result;
   }
 
+  /**
+   * Find potential transfers where traveller has to walk between stops
+   * @param from initial trip
+   * @return list of potential transfers
```

Comment:
```suggestion
   * Find potential transfers where traveller has to "walk" between stops
```
The @param do not add any information. I put walk in "walk", because it is not always walking.